### DISCUSSION TITLE
Gitlab changelog not found with sub directory

### DISCUSSION THREAD

**Discussion by pabardz** [renovatebot/renovate#18950] [Request Help]
Hey! I am trying to get changelog of my helm chart release but got an error when I tried to use helm chart in a subdirectory.

Example architecture of my chart:

```
root
...-> charts
..........-> test
.................-> all chart files (CHANGELOG.MD, etc...)
```

Basically, when my git repository architecture is with subfolders, the changelog is never found. It's found when `CHANGELOG.MD` is in root.

Error I got:

```
DEBUG: Fetching changelog: https://gitlab.random.com/myuser/test-chart/tree/master/charts/renovate (2.1.9 -> 2.2.0) (repository=myuser/renovate-2, branch=renovate/renovate-renovate-2.x)
DEBUG: GET https://gitlab.random.com/api/v4/projects/myuser%2Ftest-chart%2Ftree%2Fmaster%2Fcharts%2Frenovate/repository/tags?per_page=100 = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=38) (repository=myuser/renovate-2, branch=renovate/renovate-renovate-2.x)
DEBUG: GitLab API 404 (repository=myuser/renovate-2, branch=renovate/renovate-renovate-2.x)
       "url": "https://gitlab.random.com/api/v4/projects/myuser%2Ftest-chart%2Ftree%2Fmaster%2Fcharts%2Frenovate/repository/tags?per_page=100"
DEBUG: Failed to fetch Gitlab tags (repository=myuser/renovate-2, branch=renovate/renovate-renovate-2.x)
       "sourceRepo": "myuser/test-chart/tree/master/charts/renovate",
       "err": {
         "code": "ERR_NON_2XX_3XX_RESPONSE",
         
           }
         },
         "message": "Response code 404 (Not Found)",
         "stack": "HTTPError: Response code 404 (Not Found)\n    at Request.<anonymous> (/usr/src/app/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/34.26.2 (https://github.com/renovatebot/renovate)",
             "accept": "application/json",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://gitlab.random.com/api/v4/projects/myuser%2Ftest-chart%2Ftree%2Fmaster%2Fcharts%2Frenovate/repository/tags?per_page=100",
           "hostType": "gitlab-changelog",
           "username": "",
           "password": "",
           "method": "GET",
           "http2": false
         },
         "response": {
           "statusCode": 404,
           "statusMessage": "Not Found",
           "body": {"message": "404 Project Not Found"},
           "headers": {
             "server": "nginx",
             "date": "Wed, 16 Nov 2022 20:21:41 GMT",
             "content-type": "application/json",
             "content-length": "35",
             "connection": "close",
             "cache-control": "no-cache",
             "vary": "Origin",
             "x-content-type-options": "nosniff",
             "x-frame-options": "SAMEORIGIN",
           },
           "httpVersion": "1.1",
           "retryCount": 0
         }
       }
DEBUG: GET https://gitlab.random.com/api/v4/projects/myuser%2Ftest-chart%2Ftree%2Fmaster%2Fcharts%2Frenovate/repository/tree?per_page=100 = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=32) (repository=myuser/renovate-2, branch=renovate/renovate-renovate-2.x)
DEBUG: GitLab API 404 (repository=myuser/renovate-2, branch=renovate/renovate-renovate-2.x)
       "url": "https://gitlab.random.com/api/v4/projects/myuser%2Ftest-chart%2Ftree%2Fmaster%2Fcharts%2Frenovate/repository/tree?per_page=100"
DEBUG: Error 404 getting changelog md (repository=myuser/test-chart/tree/master/charts/renovate, branch=renovate/renovate-renovate-2.x)
       "type": "gitlab",
       "apiBaseUrl": "https://gitlab.random.com/api/v4/"
DEBUG: GET https://gitlab.random.com/api/v4/projects/myuser%2Ftest-chart%2Ftree%2Fmaster%2Fcharts%2Frenovate/releases?per_page=100 = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=48) (repository=myuser/renovate-2, branch=renovate/renovate-renovate-2.x)
DEBUG: GitLab API 404 (repository=myuser/renovate-2, branch=renovate/renovate-renovate-2.x)
       "url": "https://gitlab.random.com/api/v4/projects/myuser%2Ftest-chart%2Ftree%2Fmaster%2Fcharts%2Frenovate/releases?per_page=100"
DEBUG: getReleaseList 404 (repository=myuser/test-chart/tree/master/charts/renovate, branch=renovate/renovate-renovate-2.x)
       "type": "gitlab",
       "apiBaseUrl": "https://gitlab.random.com/api/v4/"
```

Renovate gets this information from the helm chart:

```
{
    "packageFile": "renovate/requirements.yaml",
    "deps": [
    {
        "depName": "renovate",
        "currentValue": "2.1.9",
        "registryUrls": ["https://xxxx.random.com"],
        "depIndex": 0,
        "updates": [
        {
            "bucket": "non-major",
            "newVersion": "2.2.0",
            "newValue": "2.2.0",
            "releaseTimestamp": "2022-11-16T20:11:53.000Z",
            "newMajor": 2,
            "newMinor": 2,
            "updateType": "minor",
            "branchName": "renovate/renovate-renovate-2.x"
        }
        ],
        "warnings": [],
        "versioning": "helm",
        "sourceUrl": "https://gitlab.random.com/myuser/test-chart/tree/master/charts/renovate",
        "homepage": "https://gitlab.random.com/myuser/test-chart/tree/master/charts/renovate",
        "currentVersion": "2.1.9",
        "isSingleVersion": true,
        "fixedVersion": "2.1.9"
}
          
```

So basically, the `sourceUrl` = `sourceRepo` , it generates all the 404 errors because this `sourceRepo` doesn't exist.

What am I missing?

I want to get my changelog from subfolders in a mono helm charts repository.

**Comment by pabardz** [comment]
I tested using github.com just to store the `CHANGELOG` and I updated the `home` in my Chart.yaml and it works... So there is probably an issue having changelog in subfolders with gitlab: https://github.com/pabardina/test-chart/tree/master/charts/renovate

**Reply by rarkins** [reply]
Renovate expects `sourceUrl` to point to the repo and `sourceDirectory` to be used if the changelog should be in a subdirectory. But this convention only really applies to `npm` packages for now as far as I'm aware.

**Reply by pabardz** [reply]
I don't think so, because the same setup works with GitHub and not with a self hosted gitlab. Looking at the helm sample in test file, it indeed supposes to work: 
https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/helm/common.spec.ts#L15 
https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/helm/__fixtures__/sample.yaml#L26 

So for this URL: `https://github.com/bitnami/charts/tree/master/bitnami/airflow`, Renovate splits it in `sourceUrl` and `sourceDirectory`, `${{ sourceUrl: 'https://github.com/bitnami/charts', sourceDirectory: 'bitnami/airflow' }}`,  not for GitLab though.

Looking at the code, it seems that the `findSourceUrl` function here: https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/helm/common.ts works indeed only for Github, isn't it?

**Reply by rarkins** [reply]
You're right. I had suspected there was some github-specific massaging but didn't find it. A challenge with GitLab is that we don't know for sure that it's group/project/submodule because it could also be group/subgroup/project. Is there a way which we could learn that from the datasource somehow?

**Reply by pabardz** [reply]
Hum I guess there is always the `/-/tree/branch-name` in the URL. Before `/-/tree` is the `sourceUrl` after `/-/tree/branch-name/` it's the `sourceDirectory`

For  `https://gitlab.random.com/project/test-chart/-/tree/master/charts/renovate` it would be 
`sourceUrl: https://gitlab.random.com/project/test-chart` and `sourceDirectory: charts/renovate`

And with a subgroup, it would be `https://gitlab.random.com/project/subproject/test-chart/-/tree/master/charts/renovate`  
`sourceUrl: https://gitlab.random.com/project/subproject/test-chart` and `sourceDirectory: charts/renovate`

Was that your question?

**Reply by rarkins** [reply]
Yes, I had missed that there's the `tree` part. It seems like we could split this into `sourceUrl` and `sourceDirectory`. Are you interested in submitting a PR?

**Reply by viceice** [reply]
what if the branch name contains a slash too? how we distinguish between branch name and folders

**Reply by pabardz** [reply]
@viceice Good question. The actual regex for GitHub doesn't handle that case either: https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/helm/common.ts#L6

**Reply by viceice** [reply]
@rarkins maybe we should move that parsing/ splitting to the changelogs. and we can probably use the git url parse dependency to split the url.

**Reply by rarkins** [reply]
Yeah, I think we could pass that `sourceUrl` through and then split it later

**Reply by pabardz** [reply]
Cool, you know how to fix that :) Do you need an issue or something like this for someone to start working on? Not sure I'll be useful on that one

**Reply by rarkins** [reply]
You can create a reproduction repo and then a feature request, but it doesn't mean anyone will start work on it. You are welcome to submit a PR yourself.

**Reply by colinodell** [reply]
I've opened https://github.com/renovatebot/renovate/pull/19181 which should solve this (and a related "issue" I'm having where subdirectories aren't yet supported for GitHub Enterprise deployments on custom domains like `github.some-corp.tld`)
